api/src/main/java/org/pmiops/workbench/api/DataSetController.java

```diff
@@ -300,101 +299,88 @@ private DataSetRequest generateDataSetRequestFromPreviewRequest(
-                "Timeout while querying the CDR to pull preview information.");
-          }
-          retry++;
-        } else {
-          throw ex;
-        }
+    TableResult queryResponse;
+    try {
+      String query = originalQuery.concat(" LIMIT " + NO_OF_PREVIEW_ROWS);
+
+      queryJobConfiguration = queryJobConfiguration.toBuilder().setQuery(query).build();
+
+      /* Google appengine has a 60 second timeout, we want to make sure this endpoint completes
+       * before that limit is exceeded, or we get a 500 error with the following type:
+       * com.google.apphosting.runtime.HardDeadlineExceededError
+       * See https://cloud.google.com/appengine/articles/deadlineexceedederrors for details
+       */
+      queryResponse =
+          bigQueryService.executeQuery(
+              bigQueryService.filterBigQueryConfig(queryJobConfiguration),
+              APP_ENGINE_HARD_TIMEOUT_MSEC_MINUS_FIVE_SEC);
+    } catch (Exception ex) {
+      if ((ex.getCause() != null
+          && ex.getCause().getMessage() != null
+          && ex.getCause().getMessage().contains("Read timed out"))) {
+        throw new GatewayTimeoutException(
+            "Timeout while querying the CDR to pull preview information.");
+      } else {
+        throw ex;
       }
-    } while (retry < PREVIEW_RETRY_LIMIT);
-
-    if (retry == PREVIEW_RETRY_LIMIT) {
-      throw new GatewayTimeoutException(
-          "Timeout while querying the CDR to pull preview information.");
     }
+    queryResponse
+        .getSchema()
+        .getFields()
+        .forEach(
+            fields -> {
+              valuePreviewList.add(new DataSetPreviewValueList().value(fields.getName()));
+            });
+
+    queryResponse
+        .getValues()
+        .forEach(
+            fieldValueList -> {
+              IntStream.range(0, fieldValueList.size())
+                  .forEach(
+                      columnNumber -> {
+                        try {
+                          valuePreviewList
+                              .get(columnNumber)
+                              .addQueryValueItem(
+                                  fieldValueList.get(columnNumber).getValue().toString());
+                        } catch (NullPointerException ex) {
+                          log.severe(
+                              String.format(
+                                  "Null pointer exception while retriving value for query: Column %s ",
+                                  columnNumber));
+                          valuePreviewList.get(columnNumber).addQueryValueItem("");
+                        }
+                      });
+            });
+    queryResponse
+        .getSchema()
+        .getFields()
+        .forEach(
+            fields -> {
+              DataSetPreviewValueList previewValue =
+                  valuePreviewList.stream()
+                      .filter(preview -> preview.getValue().equalsIgnoreCase(fields.getName()))
+                      .findFirst()
```

Comment:
This will blow up if it's not found, as it's calling get on an empty optional. Also, it might be faster and simpler to make a map from lowercased field name to previewValue.

Reply:
Will move to map. I can poke at the optional and see if we can get it to throw something rather than blow up though.

Reply:
I think that the filter is probably clear, and avoids an additional step, since we would need to filter the map anyway. I am moving to make it not blow up.